src/cockpit/channels/packages.py

```diff
@@ -27,50 +29,81 @@ class PackagesChannel(Channel):
-
-    def http_error(self, status, message):
+    # used to carry data forward from open to done
+    options: Optional[Dict[str, object]] = None
+
+    def http_error(self, status: int, message: str) -> None:
         template = data.read_cockpit_data_file('fail.html')
         self.send_message(status=status, reason='ERROR', headers={'Content-Type': 'text/html; charset=utf-8'})
         self.send_data(template.replace(b'@@message@@', message.encode('utf-8')))
 
-    def do_done(self):
-        assert not self.post
-        assert self.options['method'] == 'GET'
-        path = self.options['path']
+    def do_open(self, options: Dict[str, object]) -> None:
+        self.ready()
 
-        self.headers = self.options['headers']
-        self.protocol = self.headers['X-Forwarded-Proto']
-        self.host = self.headers['X-Forwarded-Host']
-        self.origin = f'{self.protocol}://{self.host}'
+        self.options = options
 
-        self.router.packages.serve_file(path, self)
-        self.done()
-        self.close()
+    def do_done(self) -> None:
+        packages: Packages = self.router.packages  # type: ignore[attr-defined]  # yes, this is evil
+        assert self.options is not None
+        options = self.options
 
-    def do_data(self, data):
-        self.post += data
+        try:
+            if options.get('method') != 'GET':
+                raise ValueError(f'Unsupported HTTP method {options["method"]}')
 
-    def do_open(self, options):
-        self.post = b''
-        self.options = options
-        self.ready()
+            path = options.get('path')
+            if not isinstance(path, str) or not path.startswith('/'):
+                raise ValueError(f'Unsupported HTTP method {options["method"]}')
+
+            headers = options.get('headers')
+            if not isinstance(headers, dict) or not all(isinstance(value, str) for value in headers.values()):
+                raise ValueError(f'Unsupported HTTP method {options["method"]}')
+
+            document = packages.load_path(path, headers)
+
+            # Note: we can't cache documents right now.  See
+            # https://github.com/cockpit-project/cockpit/issues/19071
+            # for future plans.
+            out_headers = {
+                'Cache-Control': 'no-cache, no-store',
+                'Content-Type': document.content_type,
+            }
+
+            if document.content_encoding is not None:
+                out_headers['Content-Encoding'] = document.content_encoding
+
+            if document.content_security_policy is not None:
+                policy = document.content_security_policy
+
+                # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/connect-src
+                #
+                #    Note: connect-src 'self' does not resolve to websocket
+                #    schemes in all browsers, more info in this issue.
```

Comment:
Thanks for that comment. Now I understand what triggered your recent C-S-P research :grin:

Reply:
I was actually more confused by the CORS stuff, to be honest...